### PR TITLE
Vote contract

### PR DESCRIPTION
**Pull request by carllin**
1) Create a Vote Contract
2) Move ownership of LeaderScheduler into bank
3) Change TPU/TVU pipelines to use bank for leader scheduling

### CODE REVIEW

File: src/vote_program.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    pub fn id() -> Pubkey {
+        Pubkey::new(&VOTE_PROGRAM_ID)
+    }
+
+    pub fn process_transaction(
+        tx: &Transaction,
+        instruction_index: usize,
+        leader_scheduler: &mut LeaderScheduler,
```
Comment (aeyakovenko) [inline src/vote_program.rs]: this should be context free.  just recording the vote in the Account userdata

File: src/vote_program.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct Vote {
+    /// We send some gossip specific membership information through the vote to shortcut
+    /// liveness voting
+    /// The version of the ClusterInfo struct that the last_id of this network voted with
+    pub version: u64,
+    /// The version of the ClusterInfo struct that has the same network configuration as this one
+    pub contact_info_version: u64,
+    // TODO: add signature of the state here as well
```
Comment (rob-solana) [inline src/vote_program.rs]: vote should include tick_height, not clusterinfo stuff, right?

File: src/vote_program.rs (hunk 1 above)
Comment (rob-solana) [reply src/vote_program.rs]: the implication is that the bank has a record of accounts in under the vote program, and that we can iterate over those accounts state to compute leader?

**Comment by garious** [comment]
:)

File: src/vote_program.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        output[2..=serialized_len as usize + 1].clone_from_slice(&self_serialized);
+        Ok(())
+    }
+
+    pub fn process_transaction(
+        tx: &Transaction,
+        instruction_index: usize,
+        accounts: &mut [&mut Account],
+    ) -> Result<()> {
+        if !Self::check_id(&accounts[1].program_id) {
```
Comment (aeyakovenko) [inline src/vote_program.rs]: can you document what you expect accounts 0 and 1 to be?  I was expecting only 1 account, 0, which would store the vote for this tx.

**Pull request merged**